stream.js

```diff
@@ -219,13 +227,96 @@
                 return
             }
             console.log(this.sockId,'tryClose')
-            sockets.tcp.send(this.sockId, new ArrayBuffer, callback)
+            chrome.sockets.tcp.send(this.sockId, new ArrayBuffer, callback)
         },
         cleanup: function() {
             this.writeBuffer = new WSC.Buffer
         }
     }
 
+
+     var arrayBuffer2String = function(buf) {
+		var bufView = new Uint8Array(buf);
+		var chunkSize = 65536;
+		var result = '';
+		for (var i = 0; i < bufView.length; i += chunkSize) {
+		  result += String.fromCharCode.apply(null, bufView.subarray(i, Math.min(i + chunkSize, bufView.length)));
+		}
+		return result;
+     }
+
+
+    var IOStreamTls = function(sockId, privateKey, serverCert) {
+        this.writeCallbacks = [];
+        this.readCallbacks = [];
+        var _t = this;
+
+		this.tlsServer = forge.tls.createConnection({
+		  server: true,
+		  sessionCache: {},
+		  // supported cipher suites in order of preference
+		  cipherSuites: [
+			forge.tls.CipherSuites.TLS_RSA_WITH_AES_128_CBC_SHA,
+			forge.tls.CipherSuites.TLS_RSA_WITH_AES_256_CBC_SHA],
+		  connected: function(c) {
+			//console.log('Server connected');
+			//c.prepareHeartbeatRequest('heartbeat');
+		  },
+		  verifyClient: false,
+		  getCertificate: function(c, hint) {
+			//console.log('Server getting certificate for \"' + hint[0] + '\"...');
+			return serverCert; //WSC.Tls.data.server.cert;
+		  },
+		  getPrivateKey: function(c, cert) {
+			//console.log('Server getting privateKey for \"' + cert + '\"...');
+			return privateKey;//WSC.Tls.data.server.privateKey;
+		  },
+		  tlsDataReady: function(c) {
+			// send TLS data to client
+			var cb = _t.writeCallbacks.pop() || function(){};
+			let str = c.tlsData.getBytes();
+			var b = WSC.str2ab(str);
+			//console.log('encrypt to client: ' + str);
+			if (this.connected)
+                chrome.sockets.tcp.send( _t.sockId, b, cb);
+            else
+                _t.error("tlsData on closed socket");
+		  },
```

Comment:
diff hard to read. it looks like you use tabs and not spaces :-) I don't really care though

Reply:
I am working an a patch with this and your your other suggestions.
Hopefully be done soon enough.